include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -365,72 +359,76 @@ inline void NSGA2::FastNonDominatedSort(
     i++;
     fronts.push_back(nextFront);
   }
+  // Remove the empty final set.
+  fronts.pop_back();
 }
 
 //! Check if a candidate Pareto dominates another candidate.
 template<typename MatType>
 inline bool NSGA2::Dominates(
-    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+    std::vector<arma::Col<typename MatType::elem_type> >& populationFval,
     size_t candidateP,
     size_t candidateQ)
 {
   bool allBetterOrEqual = true;
   bool atleastOneBetter = false;
-  size_t n_objectives = calculatedObjectives[0].n_elem;
+  size_t n_objectives = populationFval[0].n_elem;
 
   for (size_t i = 0; i < n_objectives; i++)
   {
     // P is worse than Q for the i-th objective function.
-    if (calculatedObjectives[candidateP](i) > calculatedObjectives[candidateQ](i))
+    if (populationFval[candidateP](i) > populationFval[candidateQ](i))
       allBetterOrEqual = false;
 
     // P is better than Q for the i-th objective function.
-    else if (calculatedObjectives[candidateP](i) < calculatedObjectives[candidateQ](i))
+    else if (populationFval[candidateP](i) < populationFval[candidateQ](i))
       atleastOneBetter = true;
   }
 
   return allBetterOrEqual && atleastOneBetter;
 }
 
 //! Assign crowding distance to the population.
-inline void NSGA2::CrowdingDistanceAssignment(const std::vector<size_t>& front,
-                                              std::vector<double>& crowdingDistance)
+template <typename MatType>
+inline void NSGA2::CrowdingDistanceAssignment(
+	const std::vector<size_t>& front,
+	std::vector<arma::Col<typename MatType::elem_type>>& populationFval,
+	std::vector<double>& crowdingDistance)
 {
-  if (front.size() > 0)
-  {
-    for (size_t elem: front)
-      crowdingDistance[elem] = 0;
+  size_t fSize = front.size();
+  std::vector<size_t> sortedIdx(fSize); // Stores the sorted indices of the fronts.
+  std::iota(sortedIdx.begin(), sortedIdx.end(), 0u);
 
-    size_t fSize = front.size();
-
-    for (size_t m = 0; m < numObjectives; m++)
+  for (size_t m = 0; m < numObjectives; m++)
+  {
+    std::vector<double> fValues{};
```

Comment:
Looks like it would be possible to use an `arma::vec` here, unless I missed something and doing so would make sorting more complex?

Reply:
Nice catch! :)